File: H_Simulator_FVM_2D/07_H_Simulator_FVM_2D/MATLAB/src/FIMPressure2D_1Phase.m
function [State] = FIMPressure2D_1Phase(Flow,Gen,State,Wells)
%% - Pressure Simulator
%Compressible 2D 1 phase pressure simulator. Uses FIM.
%Barnaby Fryer - EPFL - 29.03.17
    
    
    %Store the state variable from previous time step for use in residual
    State0 = State;
    
    %% - Solve Flow Model
    %Convergence checker
    Converged = 0;                                              %[1,1] 
    %Number of iterations
    it = 1;                                                     %[1,1]
    while Converged == 0

        %Find the non-fluid rock transmissibility
        [Trans,~,~] = Trans_Rock(Flow,Gen,State);               %2x[Nx,1]
        
        %Build residual
        [Res] =...
            BuildResidual(Flow,Gen,State,State0,Trans,Wells);   %[N,1]
        
        %Build Jacobian
        [Jac] = BuildJacobian(Flow,Gen,State,Trans,Wells);      %[N,N]
        
        %Solve
        x = -Jac\Res;                                           %[N,1]
        
        %Update Solution
        State.P = State.P + x;                                  %[N,1]
        
        %Build residual
        [Res] =...
            BuildResidual(Flow,Gen,State,State0,Trans,Wells);   %[N,1]
        
        %Check Convergence
        %Computes infinite norm of residual
        Norm = norm(Res,inf);                                   %[1,1]                                               
        if Norm < Gen.tol
            %Set to converged
            Converged = 1;                                      %[1,1]           
        else
            %Update iteration tracker
            it = it + 1;                                        %[1,1]                 
        end  
    end

    
    %Find fluid flux
    %Upwind
    [A] = Upwind(Gen,State,Trans);                              %[N,N][N,N] 
    %Fluid tansmissivity
    [FTrans,~] = Trans_Fluid(A,Flow,Gen,State);                 %[N,N]
    %Find source terms
    [Q,~] = Add_Wells(Flow,State,Wells);                        %[N,1]
    %Find convection into cell
    State.flux = -FTrans*State.P + Q;                           %[N,1]
    %Calculate perm and derivative
    [State,~] = Perm(Flow,State);                	            %[N,1]
    %Calculate porosity and derivative
    [State.phi,~] = PhiCalc(Flow,State);                	    %[N,1]

end

File: HM_Simulator_FEM_FVM_2D/09_HM_Simulator_FEM_FVM_2D/Python/src/src_flow/FIMPressure_2D_1Phase.py
import numpy as np

from src.src_flow.phiCalc import phiCalc
from src.src_flow.trans_rock import trans_rock
from src.src_flow.trans_fluid import trans_fluid
from src.src_flow.upwind import upwind
from src.src_flow.perm import perm
from src.src_flow.build_jacobian import build_jacobian
from src.src_flow.build_residual import build_residual
from src.src_flow.PStrainUpdate import PStrainUpdate

from scipy.sparse.linalg import spsolve


def FIMPressure_2D_1Phase(Flow,Gen,State,State0,State_phi,Wells):

    # =============================================================================
    # Update pressure based on volumetric strain from mechanics model, conserve mass
    # =============================================================================

    State = PStrainUpdate(Flow,Gen,State,State_phi)

    # =============================================================================
    # Solve flow model
    # =============================================================================
    #Convergence checker
    Converged = 0
    #Number of iterations
    it = 1
    while Converged != 1:

        # Find the non-fluid rock transmissibility
        trans,_,_ = trans_rock(Flow, Gen, State)

        #Build residual
        res = build_residual(Flow,Gen,State,State0,trans,Wells)
        #Build Jacobian
        jac = build_jacobian(Flow,Gen,State,trans,Wells)

        #Solve
        x = spsolve(jac, -res)

        #Update solution
        State["P"] = State["P"] + x

        # Build residual
        res = build_residual(Flow, Gen, State, State0, trans,Wells)

        #Check convergence
        #Computes infinite norm of residual
        norm = np.linalg.norm(res, ord=np.inf)
        if norm < Gen["tol"]:
            #Set to converged
            Converged = 1
        else:
            #Update iteration tracker
            it += 1

    # =============================================================================
    # Find fluid flux
    # =============================================================================

    #Upwind
    A = upwind(Gen,State,trans)
    #Fluid transmissivity
    Ftrans,_ = trans_fluid(A,Flow,Gen,State)
    #Fluid flux into cell
    State["flux"] = -Ftrans @ State["P"]
    #Find permeability
    State["kx"],State["ky"],_,_,_,_ = perm(Flow,State)
    #Find porosity
    State["phi"],_ = phiCalc(Flow,State)

    return State